std/traits.d

```diff
@@ -6729,3 +6730,100 @@ unittest
     static assert(getSymbolsByUDA!(C, UDA)[0].stringof == "C");
     static assert(getSymbolsByUDA!(C, UDA)[1].stringof == "d");
 }
+
+/**
+ * A static assertion that a type satisfies a given template constraint.
+ * It can be used as a $(LINK2 ../attribute.html#uda, UDA) or
+ * in a $(D static assert) to make sure that a type conforms
+ * to the compile-time interface the user expects it to.
+ * The difference between using $(D models) and a simple static assert
+ * with the template contraint is that $(D models) will instantiate the
+ * failing code when the constraint is not satisfied,
+ * yielding compiler error messages to aid the user.
+ *
+ * The template contraint predicate must start with the word $(D is)
+ * (e.g. $(D isInputRange)) and an associated template function
+ * with the "is" replaced by "check" should exist (e.g. $(D checkInputRange))
+ * and be defined in the same module.
+ */
+template models(T, alias P, A...)
+{
+    static assert(P.stringof[0..2] == "is",
+                  P.stringof ~ " does not begin with 'is', which is require by `models`");
+
+    static if(P!(T, A))
+    {
+        bool models()
+        {
+            return true;
+        }
+    }
+    else
+    {
+        bool models()
+        {
+            import std.algorithm;
+            enum openParenIndex = P.stringof.countUntil("(");
+            //2 to get rid of "is"
+            enum untilOpenParen = P.stringof[2 .. openParenIndex];
+            enum checkName = "check" ~ untilOpenParen;
+            enum mixinStr = checkName ~ "!(T, A);";
+            mixin("import " ~ moduleName!(P) ~ ";"); //make it visible first
```

Comment:
Doesn't this entail massive code bloat as you're important at least one module per type?